sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -237,21 +237,24 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           new Path(metastoreRelation.catalogTable.storage.locationUri.get),
           partitionSpec)
 
-        val inferredSchema = if (fileType.equals("parquet")) {
-          val inferredSchema =
-            defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
-          inferredSchema.map { inferred =>
-            ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, inferred)
+        val inferredSchema =
+          defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
+        val schema = if (fileType == "parquet") {
+          // For Parquet, get correct schema by merging Metastore schema data types
+          // and Parquet schema field names.
+          inferredSchema.map { schema =>
+            ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, schema)
           }.getOrElse(metastoreSchema)
         } else {
-          defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles()).get
+          // For others (e.g orc), fall back to metastore schema if needed.
```

Comment:
If the fileType is not parquet, the only possible type is orc. To ensure this is always true, we should check `fileType == "orc"`.